cranelift/codegen/src/machinst/buffer.rs

```diff
@@ -824,6 +825,38 @@ impl<I: VCodeInst> MachBuffer<I> {
                 // (i.e., an infinite loop).
                 if self.resolve_label_offset(b.target) != b.start {
                     let redirected = b.labels_at_this_branch.len();
+                    // We cannot perform this aliasing if the target of this
+                    // branch ultimately aliases back here; if so, we need to
+                    // keep this branch, so break out of the loop entirely.
+                    //
+                    // Note that this check is what prevents cycles from forming in
+                    // `self.label_aliases`. To see why, consider an arbitrary start
+                    // state:
+                    //
+                    // label_aliases[L1] = L2, label_aliases[L2] = L3, ..., up to
+                    // Ln, which is not aliased.
+                    //
+                    // We would create a cycle if we assigned label_aliases[Ln]
+                    // = L1.  Note that the below assignment is the only write
+                    // to label_aliases.
+                    //
+                    // By our other invariants, we have that Ln (`l` below)
+                    // resolves to the offset `b.start`, because it is in the
+                    // set `b.labels_at_this_branch`.
+                    //
+                    // If L1 were already aliased, through some arbitrarily deep
+                    // chain, to Ln, then it must also resolve to this offset
+                    // `b.start`.
+                    //
+                    // By checking the resolution of `L1` against this offset,
+                    // and aborting this branch-simplification if they are
+                    // equal, we prevent the below assignment from ever creating
+                    // a cycle.
+                    if self.resolve_label_offset(b.target) == b.start {
```

Comment:
Unless I'm missing something, the exact opposite of this condition is checked right above on line 826, with no apparent side-effects in between, so this is unreachable code, and the existing code already does the right thing. In this case, moving the comment up could be useful.

Reply:
Excellent point, I had missed this. I made a slight change to the existing `if`: we need an `else` that stops the branch-folding loop entirely, to preserve invariants.